propagation.go

```diff
@@ -41,71 +45,112 @@ var (
+	//
+	// For Tracer.Inject(): the carrier must be an `io.Writer`.
 	//
-	// The `carrier` for injection and extraction must be a
-	// `*SplitBinaryCarrier` instance.
-	SplitBinary BuiltinFormat = iota
+	// For Tracer.Join(): the carrier must be an `io.Reader`.
+	Binary BuiltinFormat = iota
 
-	// SplitText encodes the Span in a SplitTextCarrier instance.
+	// TextMap encodes the Span in a TextMapCarrier instance.
+	//
+	// For Tracer.Inject(): the carrier must be a `TextMapWriter`.
+	//
+	// For Tracer.Join(): the carrier must be a `TextMapReader`.
+	//
+	// See HTTPHeaderTextMapCarrier for an implementation of both TextMapWriter
+	// and TextMapReader that defers to an http.Header instance for storage.
+	// For example, Inject():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    err := span.Tracer().Inject(span, TextMap, carrier)
 	//
-	// The `carrier` for injection and extraction must be a `*SplitTextCarrier`
-	// instance.
+	// Or Join():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    span, err := tracer.Join("opName", TextMap, carrier)
+	//
+	TextMap
+
+	// SplitBinary is DEPRECATED
+	SplitBinary
+	// SplitText is DEPRECATED
 	SplitText
+)
+
+// TextMapWriter is the Inject() carrier for the TextMap builtin format. With
+// it, the caller can encode a Span for propagation as entries in a multimap of
+// unicode strings.
+type TextMapWriter interface {
+	// Add a key:value pair to the carrier. Multiple values may be added for a
+	// single (repeated) key.
+	Add(key, val string)
+}
 
-	// GoHTTPHeader encodes the Span into a Go http.Header instance (both the
-	// tracer state and any baggage).
+// TextMapWriter is the Join() carrier for the TextMap builtin format. With it,
+// the caller can decode a propagated Span as entries in a multimap of unicode
+// strings.
+type TextMapReader interface {
+	// ReadAll returns TextMap contents via repeated calls to the `handler`
+	// function. If any call to `handler` returns a non-nil error, ReadAll
+	// terminates and returns that error.
 	//
-	// The `carrier` for both injection and extraction must be an http.Header
-	// instance.
-	GoHTTPHeader
-)
+	// NOTE: A single `key` may appear in multiple calls to `handler` for a
+	// single `ReadAllEntries` invocation.
+	ReadAll(handler func(key, val string) error) error
+}
 
-// SplitTextCarrier breaks a propagated Span into two pieces.
+// HTTPHeaderTextMapCarrier satisfies both TextMapWriter and TextMapReader.
 //
-// The Span is separated in this way for a variety of reasons; the most
-// important is to give OpenTracing users a portable way to opt out of
-// Baggage propagation entirely if they deem it a stability risk.
-//
-// It is legal to provide one or both maps as `nil`; they will be created
-// as needed. If non-nil maps are provided, they will be used without
-// clearing them out on injection.
+// NOTE: All `key` parameters to Add() and the ReadAllEntries() handler func
```

Comment:
I can't decide whether it's fair to say that `HTTPHeaderTextMapCarrier` actually satisfies `TextMap(Writer|Reader)` given the string casing issues here. Oh well.